lib/inspec/plugins/resource.rb

```diff
@@ -65,6 +71,14 @@ def skip_resource(message)
           @resource_skipped = message
         end
 
+        def resource_failed
+          @resource_failed if defined?(@resource_failed)
```

Comment:
If `@resource_failed` is never assigned a value, it will return `nil` by default. Do we need this `defined?` check?

Reply:
I agree with you, I honestly just repeated what was there. I started down the path of correcting it and noticed that quite a few unrelated tests checked the value of `resource_skipped` and would need to be changed to `resource_skipped_message`.

The real worry here is that some of these tests are Unit tests for resources. Which got me thinking, if I modify `resource_skipped` to be a `?` and `_message` would that be a breaking change for some?

If so, should I:
  - Fix `resource_failed` only and have it be different than `resource_skipped`? (and put in an issue for InSpec 2.0 to correct the difference)
  - Just leave it as is and put in an issue for InSpec 2.0 to fix both?

Thoughts?

Reply:
Just because something was there doesn't mean it's the right thing to do. :) We should always look for improvements while mucking around in the codebase.

The question to ask is: "are these methods part of an expected public API?" Is `resource_failed` used anywhere other than this class and its unit tests? Do the consumers of this method extend beyond the InSpec codebase (i.e. can a user call these methods from within their tests, and is it something we support)? And, if the answer is yes to either of these, can we do something to transition to what we want it to be?

If you agree that having `?` methods is a better fit here, then I'd like to encourage you to think about ways to get there without introducing breaking changes.  For example:

 - add the `?` methods
 - add the new `_message` methods
 - create method aliases for the non-`?` methods to the `_message` methods

I think you'll find it you do that, you gain a more idiomatic implementation while maintaining backward compatibilities (and we should add code comments explaining why the method aliases exist) and then we can remove those aliases in the future.

Reply:
Excellent suggestions! I didn't think about an alias method. I pushed up another commit implementing what you recommended.

It also modifies other files that used `resource_skipped` in order to reduce work in the future when we deprecate. Let me know if I need to remove those changes.